backend/src/hatchling/utils/context.py

```diff
@@ -71,73 +74,73 @@ def __format_env(self, value, data):
 
 
 class Context:
-    def __init__(self, root):
+    def __init__(self, root: str) -> None:
         self.__root = str(root)
 
         # Allow callers to define their own formatters with precedence
-        self.__formatters = ChainMap()
-        self.__configured_contexts = set()
+        self.__formatters: ChainMap = ChainMap()
+        self.__configured_contexts: set[str] = set()
         self.__formatter = ContextStringFormatter(self.__formatters)
 
         self.add_context(DefaultContextFormatter(self.__root))
 
-    def format(self, *args, **kwargs):
+    def format(self, *args: Any, **kwargs: Any) -> str:
         return self.__formatter.format(*args, **kwargs)
 
-    def add_context(self, context):
+    def add_context(self, context: DefaultContextFormatter) -> None:
         if context.CONTEXT_NAME in self.__configured_contexts:
             return
 
         self.__add_formatters(context.get_formatters())
         self.__configured_contexts.add(context.CONTEXT_NAME)
 
     @contextmanager
-    def apply_context(self, context):
+    def apply_context(self, context: DefaultContextFormatter) -> Iterator:
         self.__add_formatters(context.get_formatters())
         try:
             yield
         finally:
             self.__remove_formatters()
 
-    def __add_formatters(self, formatters):
+    def __add_formatters(self, formatters: MutableMapping) -> None:
         return self.__formatters.maps.insert(0, formatters)
 
-    def __remove_formatters(self):
+    def __remove_formatters(self) -> None:
         if len(self.__formatters.maps) > 1:
             self.__formatters.maps.pop(0)
 
 
 class ContextStringFormatter(string.Formatter):
-    def __init__(self, formatters):
+    def __init__(self, formatters: ChainMap) -> None:
         super().__init__()
 
         self.__formatters = formatters
 
-    def vformat(self, format_string, args, kwargs):
+    def vformat(self, format_string: str, args: Sequence[Any], kwargs: Mapping[str, Any]) -> str:
         # We override to increase the recursion limit from 2 to 10
-        used_args = set()
-        result, _ = self._vformat(format_string, args, kwargs, used_args, 10)
-        self.check_unused_args(used_args, args, kwargs)
+        used_args: set[int | str] = set()
+        result, _ = self._vformat(format_string, args, kwargs, used_args, 10)  # type: ignore[attr-defined]
```

Comment:
`typeshed` no stub for `_vformat` method, https://github.com/python/typeshed/pull/9228